modules/backend/behaviors/ListController.php

```diff
@@ -165,7 +165,8 @@ public function makeList($definition = null)
         /*
          * List Widget with extensibility
          */
-        $widget = $this->makeWidget('Backend\Widgets\Lists', $columnConfig);
+        /** @var \Backend\Widgets\Lists $widget */
```

Comment:
Why is this required if the class reference is used to make the widget?

Reply:
Because the makeWidget method has as return typehint WidgetBase which doesn't have the methods that are used in the code below it. If we remove this we get many method does not exist errors by a decent IDE.

Reply:
There's a fine line between what the IDE says and common sense by the developer. There's very little reason in the first place for someone to be digging around in this method, and the makeList method already tells the IDE what widget is being returned.

Reply:
In my opinion good readable codebase should be the goal of every project. This includes the readability by the ide. Why would you want a file with 10 lines of errors by the ide or analyzer tool like phpstan. Phpstan also reads these kind of docblocks if I recall correctly. Coming from laravel I find it somewhat shocking to look into the octoberms codebase. So I'm trying to help you guys fixing some things up by making the code readable for an ide and a person.

Reply:
I'm going to defer to @daftspunk's opinion on this one, I personally don't like having `@var` comments littered all over the codebase since they provide little benefit in my eyes. I appreciate your effort in cleaning things up a bit though and your patience as we decide how we want to handle these sort of additions.

Reply:
>  There's very little reason in the first place for someone to be digging around in this method

My personal opinion:
I find this odd. We should be encouraging people to get to know the system to such a degree that they end up digging around this method (any method like that really). It means they're looking to understand OC as a system. For a project that (I think) would like to have more contributions from the community it'd be great if they understand most of the system.

Having that said, allowing these kind of docblock lines will help most of the major IDE's understand exactly what the code is doing and enable functionality like jump to definition. It would help people understand the system at a faster pace.

![list widget setSearchTerm](https://user-images.githubusercontent.com/3026259/77200153-4cb45680-6aea-11ea-8c2d-61836e39b4f4.png)

As an example, this is what my IDE currently knows about the method `setSearchTerm` (which is pretty much just listing any class having the method). While we could tell the IDE it's the first one.

Reply:
Is there no way to train the makeWidget method to have a return type of the first argument?

Reply:
@daftspunk nog that I know of, since we don't have generics in PHP.

Reply:
If you change it from returning WidgetBase to a generic object, will it stop the errors?

The proposed solution is a WET monkey patch

Reply:
@daftspunk No that wouldn't fix the errors. It's good that makeWidget returns WidgetBase, so I wouldn't change that.

Reply:
It's not returning a WidgetBase. It's returning a generic object that inherits WidgetBase. Unfortunately the IDE is not smart enough to figure this out. I wonder how it would handle magic methods?

The issue is not everyone uses an IDE and this line of code makes no sense, and cannot be verified to be correct, so we will constantly face this problem of "incorrectness". You must treat the problem and not the symptom. Someh

Reply:
@daftspunk Phpstorm isn't a fan of magic method either 😄 . I don't know how you want me to fix it. If we remove the docblock my IDE warns about this, and not the accessed methods I talked about earlier (sorry for that).
![afbeelding](https://user-images.githubusercontent.com/15017400/77205475-3bbd1280-6af5-11ea-896c-6155dffa6b60.png)
Changing the return type of this method to WidgetBase also feels wrong, then I'm leaning more towards leaving this docblock out and accept the warning of Phpstorm.

Reply:
I understand. The IDE helpers can be  overly strict which can stifle innovation; the things that make PHP special and October possible in the first place

My preference would be to make the return value generic and sacrifice the intellisense, so we don't keep revisiting this issue. But, since you are leading this initiative as an IDE user, we can follow your preference

It is a shame but hopefully the IDE will improve to support more flexibility in dynamic languages, or maybe we will think of a workable solution in the meantime, one that addresses the problem at its source

Reply:
@daftspunk what do you think about changing string|void to string|null as I talked about in the pull request comment.

Reply:
@daftspunk another possibility is to change the return type of `\Backend\Widgets\Lists` to `\Backend\Classes\WidgetBase|\Backend\Widgets\Lists`. Then we havo no errors in the code

Reply:
@jacobdekeizer that's worse, because now we have to start listing every single possible implementation of WidgetBase.

Reply:
@LukeTowers @daftspunk I'm still against removing @var tbh. Why can't we help the ide out so we have good autocompletion. I saw in pr #4960 someone also used @var notation to help the ide out. I don't think it's bad practice, it's the nature of php that we have to do this.

Reply:
As discussed, the method returns `object|\Backend\Classes\WidgetBase`. We want to resolve IDE errors as first priority, autocomplete is secondary

Reply:
The @var notation fixes both issues. We fix the ide errors and it returns the correct type.

Reply:
> The issue is not everyone uses an IDE and this line of code makes no sense, and cannot be verified to be correct, so we will constantly face this problem of "incorrectness". You must treat the cause and not the symptom

Reply:
@daftspunk if someone doesn't use an ide, then all the code is dus plain text. What difference does that one line then make? Instead all the ide users get good typehinting and autocompletion. We can return WidgetBase, but in my opinion adding `@var` and returning the  Lists object would be so much better and the only right thing to do. But if you guys are so against that one line then I will remove it.

Reply:
Exactly, it is an opinion that says "Use PHPStorm/IDE". It's not just one line. It's a modification of existing architecture where that line must be introduced everywhere. Even in your own Pull Request, it is used some places and not others, how to verify this? How to maintain this?

Thank you for your understanding